tests/test_filtering.py

```diff
@@ -0,0 +1,251 @@
+
+    # testing a single pixel source
+    input_box_centre = np.zeros((up.HII_DIM,) * 3, dtype="f4")
+    input_box_centre[up.HII_DIM // 2, up.HII_DIM // 2, up.HII_DIM // 2] = 1.0
+    output_box_centre = np.zeros((up.HII_DIM,) * 3, dtype="f8")
+    # use MFP=20 for the exp filter, use a 3 cell shell for the annular filter
+    if filter_flag == 3:
+        R_param = 20
+    elif filter_flag == 4:
+        R_param = max(R - 3 * (up.BOX_LEN / up.HII_DIM), 0)
+    else:
+        R_param = 0
+
+    lib.test_filter(
+        up(),
+        cp(),
+        ap(),
+        fo(),
+        ffi.cast("float *", input_box_centre.ctypes.data),
+        R,
+        R_param,
+        filter_flag,
+        ffi.cast("double *", output_box_centre.ctypes.data),
+    )
+
+    # expected outputs given in cell units
+    R_cells = R / up.BOX_LEN * up.HII_DIM
+    Rp_cells = R_param / up.BOX_LEN * up.HII_DIM
+    r_from_centre = np.linalg.norm(
+        np.mgrid[0 : up.HII_DIM, 0 : up.HII_DIM, 0 : up.HII_DIM]
+        - np.array(
+            [
+                up.HII_DIM // 2,
+                up.HII_DIM // 2,
+                up.HII_DIM // 2,
+            ]
+        )[:, None, None, None],
+        axis=0,
+    )
+    # average value of R in a sphere with same volume of the cell
+    r_from_centre[up.HII_DIM // 2, up.HII_DIM // 2, up.HII_DIM // 2] = (
+        np.pi * (0.62035) ** 4
+    )
+    exp_output_centre = get_expected_output_centre(
+        r_from_centre, R_cells, Rp_cells, filter_flag
+    )
+
+    if plt == mpl.pyplot:
+        r_bins = np.linspace(0, (up.HII_DIM / 2 * np.sqrt(3)), num=32)
+        r_cen = (r_bins[1:] + r_bins[:-1]) / 2
+        binned_truth_centre = get_expected_output_centre(
+            r_cen, R_cells, Rp_cells, filter_flag
+        )
+        filter_plot(
+            inputs=[
+                input_box_centre,
+            ],
+            outputs=[
+                output_box_centre,
+            ],
+            binned_truths=[
+                binned_truth_centre,
+            ],
+            truths=[
+                exp_output_centre,
+            ],
+            r_bins=r_bins,
+            r_grid=r_from_centre,
+            slice_index=up.HII_DIM // 2,
+            slice_axis=0,
+            plt=plt,
+        )
+
+    # All filters should be normalised aside from the exp filter
+    if filter_flag == 3:
+        # ratio of exponential and sphere volume integrals
+        norm_factor = (
+            2 * R_param**3
+            - R_param * np.exp(-R / R_param) * (2 * R_param**2 + 2 * R_param * R + R**2)
+        ) / (3 * R**3)
+    else:
+        norm_factor = 1
+    # firstly, make sure the filters are normalised
+    np.testing.assert_allclose(
+        input_box_centre.sum() * norm_factor, output_box_centre.sum(), rtol=1e-4
+    )
+
+    # then make sure we get the right shapes (more lenient for aliasing)
+    np.testing.assert_allclose(
+        output_box_centre, exp_output_centre, rtol=RELATIVE_TOLERANCE
+    )
```

Comment:
I think using an absolute tolerance makes more sense here (at least for the normalized filters), because you have a bunch of cells that are really close to zero.